activerecord/lib/active_record/validations/uniqueness.rb

```diff
@@ -85,12 +85,14 @@ def build_relation(klass, attribute, value)
 
       def scope_relation(record, relation)
         Array(options[:scope]).each do |scope_item|
-          scope_value = if record.class._reflect_on_association(scope_item)
-            record.association(scope_item).reader
+          scope = {}
+          if reflection = record.class._reflect_on_association(scope_item)
+            scope[reflection.foreign_key]  = record[reflection.foreign_key]
+            scope[reflection.foreign_type] = record[reflection.foreign_type] if reflection.polymorphic?
           else
-            record._read_attribute(scope_item)
+            scope[scope_item] = record._read_attribute(scope_item)
           end
-          relation = relation.where(scope_item => scope_value)
+          relation.where!(scope)
```

Comment:
This changes the argument passed. Also why are you using both `[]` and `_read_attribute`? What about this?

```ruby
def scope_relation(record, relation)
  attributes = Array(options[:scope]).flat_map do |scope_item|
    if (reflection = record.class._reflect_on_association(scope_item))
      if reflection.polymorphic?
        [reflection.foreign_key, reflection.foreign_type]
      else
        [reflection.foreign_key]
      end
    else
      [scope_item]
    end
  end
  attributes.inject(relation) do |product_relation, attribute|
    product_relation.where(attribute => record._read_attribute(attribute)
  end
end